### PR TITLE
fix: fix permit signer

### PR DESCRIPTION
**Pull request by rndquu**
Resolves https://github.com/ubiquity/pay.ubq.fi/issues/323

TLDR; [This](https://github.com/ubiquity-os/permit-generation/pull/70) PR broke permit signature in [this](https://github.com/ubiquity-os/permit-generation/pull/70/files#diff-a85780b2806b3d56cfa46cb3ae9e404ee06a3c1221d0508ef4c8bee91048a784R95) line.

**Root cause**

Check [this](https://github.com/ubiquity/audit.ubq.fi/issues/15#issuecomment-2421534945) buggy permit where permit owner is `0x9051eDa96dB419c967189F4Ac303a290F3327680` but recovered signer from signature is `0x51afbf762cdedc2683afecc2ccdd93da6a071ff8` while the expected behavior for both of those values to be equal. The root cause is that [here](https://github.com/ubiquity-os/permit-generation/pull/70/files#diff-a85780b2806b3d56cfa46cb3ae9e404ee06a3c1221d0508ef4c8bee91048a784R95) `domain.version` was added to the signed data while original permit2's EIP712 contract [verifies](https://github.com/Uniswap/permit2/blob/cc56ad0f3439c502c246fc5cfcc3db92bb8b7219/src/EIP712.sol#L34) only: `domain.name`, `domain.chainId` and `domain.verifyingContract`. Thus adding a new field to signed data made a signature created by the `ubiquity-os/permit-generation` package not matching the one actually [verified](https://github.com/Uniswap/permit2/blob/cc56ad0f3439c502c246fc5cfcc3db92bb8b7219/src/SignatureTransfer.sol#L65).

This PR:
- removes `domain.version` from signed data (and keeps only `domain.name`, `domain.chainId` and `domain.verifyingContract` originally produced by [SignatureTransfer.getPermitData()](https://github.com/rndquu/permit-generation/blob/e4e8eaef1d065cffdcb379b001c62d0fa971010e/src/handlers/generate-erc20-permit.ts#L91))
- adds a unit test so that possible issues with permit signatures are spotted easily

The PR that broke permit signatures was introduced in `ubiquity-os/permit-generation v2.0.0` and `ubiquity-os-marketplace/text-conversation-rewards` started using that version in `v1.5.1` on 12 October. So all permits generated since the 12 October are not claimable by contributors (and probably should be regenerated on demand once a new version of `ubiquity-os/permit-generation` is released).

P.S. The error could not be catched locally at `pay.ubq.fi` because `pay.ubq.fi` test helper script [generates](https://github.com/ubiquity/pay.ubq.fi/blob/625a74be0adbb282cc46901ac2df945fcfc3fa6e/scripts/typescript/generate-erc20-permit-url.ts#L42) a permit signature without the help of the `ubiquity-os/permit-generation` plugin.

### CODE REVIEW

**Comment by rndquu** [comment]
Somehow I don't have permissions to ask for a code review although my account is a member of the `ubiquity-os` organization. 

Pls review @gentlementlegen @whilefoo @0x4007

**Review by gentlementlegen — COMMENTED** [review]
@rndquu Thanks a lot for looking into this. After this is deployed on NPM, we should update `text-conversation-reward` package version and regenerate the permits then.
I will try to test this locally.

**Comment by rndquu** [comment]
> we should update `text-conversation-reward` package version

As far as I understand because of `^` [here](https://github.com/ubiquity-os-marketplace/text-conversation-rewards/blob/5da20ef41fcf2d54c7951f0f2882da35aefd5be6/package.json#L38) it will automatically fetch the next `v2.0.5`

**Comment by gentlementlegen** [comment]
@rndquu It should indeed, we still need to rebuild it because we run a [compiled version](https://github.com/ubiquity-os-marketplace/text-conversation-rewards/tree/development/dist).

**Comment by gentlementlegen** [comment]
@rndquu Somehow having a har time testing this, sorry for the delay. I followed the `README` to create a proper key and encrypted EVM to run it, but I run into the following issue:
- you wrote that the key should not start with `0x` but I get this error `Error: Failed to instantiate wallet: Error: invalid hexlify value (argument="value", value="d9530<rest of the wallet hex>", code=INVALID_ARGUMENT, version=bytes/5.7.0)`
- when adding `0x` I get `Error: Failed to instantiate wallet: Error: invalid private key (argument="privateKey", value="[[ REDACTED ]]", code=INVALID_ARGUMENT, version=signing-key/5.7.0)`

I'll keep debugging but you might find out faster than me.

**Comment by rndquu** [comment]
> @rndquu Somehow having a har time testing this, sorry for the delay. I followed the `README` to create a proper key and encrypted EVM to run it, but I run into the following issue:
> 
> * you wrote that the key should not start with `0x` but I get this error `Error: Failed to instantiate wallet: Error: invalid hexlify value (argument="value", value="d9530<rest of the wallet hex>", code=INVALID_ARGUMENT, version=bytes/5.7.0)`
> * when adding `0x` I get `Error: Failed to instantiate wallet: Error: invalid private key (argument="privateKey", value="[[ REDACTED ]]", code=INVALID_ARGUMENT, version=signing-key/5.7.0)`
> 
> I'll keep debugging but you might find out faster than me.

I suppose since `ethersjs` was downgraded from `v6` to `v5` then the `0x` prefix is required again.

> Error: Failed to instantiate wallet: Error: invalid private key

Hard to say, you should use the format `ETHEREUM_PRIVATE_KEY:GITHUB_ORGANIZATION_ID`.

[This](https://github.com/ubiquity-os/permit-generation/blob/e4e8eaef1d065cffdcb379b001c62d0fa971010e/tests/generate-erc20-permit.test.ts#L64-L86) test actually QAs this PR pretty much. All you have to check is that [this](https://github.com/ubiquity-os/permit-generation/blob/e4e8eaef1d065cffdcb379b001c62d0fa971010e/tests/generate-erc20-permit.test.ts#L79) signature was created by [this](https://github.com/ubiquity-os/permit-generation/blob/e4e8eaef1d065cffdcb379b001c62d0fa971010e/tests/generate-erc20-permit.test.ts#L78) owner.

**Comment by gentlementlegen** [comment]
@rndquu Long story short, I had `yarn link` on `rpc-handler` into `permit-generation` that I was linking against `conversation-rewards` which was using broken code. I runs great and I could generate a [permit](https://pay.ubq.fi?claim=W3sidHlwZSI6ImVyYzIwLXBlcm1pdCIsInBlcm1pdCI6eyJwZXJtaXR0ZWQiOnsidG9rZW4iOiIweGU5MUQxNTNFMGI0MTUxOEEyQ2U4RGQzRDc5NDRGYTg2MzQ2M2E5N2QiLCJhbW91bnQiOiIyMTAwMDAwMDAwMDAwMDAwMDAwIn0sIm5vbmNlIjoiODMwNzY0Mzc0NDY0OTk1ODkzMDQ2MTExMjg5NjM5MTY3MTAxMDY4ODYxMDI0MzYwMjM4ODE1MjAwNTgzNDY4MDA1Nzg3NTQ0MDE3NTUiLCJkZWFkbGluZSI6IjExNTc5MjA4OTIzNzMxNjE5NTQyMzU3MDk4NTAwODY4NzkwNzg1MzI2OTk4NDY2NTY0MDU2NDAzOTQ1NzU4NDAwNzkxMzEyOTYzOTkzNSJ9LCJ0cmFuc2ZlckRldGFpbHMiOnsidG8iOiIweDBmQzFiOTA5YmE5MjY1QTg0NmI4MkNGNENFMzUyZmMzZTdFZUIyRUQiLCJyZXF1ZXN0ZWRBbW91bnQiOiIyMTAwMDAwMDAwMDAwMDAwMDAwIn0sIm93bmVyIjoiMHhkOTUzMEYzZmJCRWExMWJlRDAxREMwOUU3OTMxOGYyZjIwMjIzNzE2Iiwic2lnbmF0dXJlIjoiMHhmZGM5NjBmZDkxNTEwMjU0ZWVhMTVkZTVmNmFiYmQ4N2YzZjdkZGRkY2E3ZGRmMjVmOTExODg3ODdjZTgzNzRjMDFiODFmYjYzZDc5ZmY4NjhhMjE5ZDIwM2I0YjMyYjNmOTFkMjIwNzg0OGRjMGYzN2I1ZDJmY2IwOGE1ZmMzYzFiIiwibmV0d29ya0lkIjoxMDB9XQ==).

**Review by gentlementlegen — APPROVED** [review]

**Review by 0x4007 — APPROVED** [review]

**Review by whilefoo — APPROVED** [review]
